Use always new appender

Zelldon opened the pull request:
## Description

 * Previous if the leader close transition took to long and the broker become leader again it could happen that the new appender was used by the old log storage appender. This is now prevented by not using a supplier. It requests the appender once on leader transition and always uses the same object.

@deepthidevaki you mentioned that you want to prevent stuff based on different term. I think it is already something checked in the ZeebePartition regarding that. If this is not enough for your we can create a follow up issue.

<!-- Please explain the changes you made here. -->

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #6346

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

deepthidevaki commented on broker/src/main/java/io/zeebe/broker/system/partitions/impl/steps/AtomixLogStoragePartitionStep.java:
```diff
@@ -0,0 +1,48 @@
+
+    final var appenderOptional = server.getAppender();
+    appenderOptional.ifPresentOrElse(
+        logAppender -> {
```
What if this `logAppender` is already from a newer term? This is not checked else where, right? So, in theory, it is still possible for a LogstorageAppender from an older term to write to a LeaderRole at newer term.

deepthidevaki replied:
That check doesn't prevent this, right? It is only checking if there was an error on notifying listeners and there is already a new role, then ignore the error. Because the new transition might succeed.

Do you think it is going to be a big change to make it a followup issue? IMO without checking the term, the bug is not really fixed.

Zelldon replied:
Lets talk about via zoom

Zelldon commented:
Hey @deepthidevaki I introduced the new term parameter. I tried several approaches to only add it on the leader transition etc (with having a factory) but made it more complicated then necessary (and was not easily solvable with the currentTransition etc.) This is the reason why I added the term on both leader and follower. For the inactive it was not possible, since it can also be called from the RaftFailureListener, which is why I used a "magic number" for the inactive role. We can discuss this ofc.

As discussed I added only a check for the term in the AtomixLogStoragePartitionStep and not on the PartitionTransitionImpl, which we could later at as an optimization.

deepthidevaki reviewed (approved):
Thanks @Zelldon

Zelldon commented:
bors r+

Zelldon commented:
@deepthidevaki we probably want to backport this as well right?

deepthidevaki commented:
Yes. We should backport.
